traffic_ops/traffic_ops_golang/deliveryservices_keys.go

```diff
@@ -34,10 +34,63 @@ import (
 	"github.com/apache/incubator-trafficcontrol/lib/go-log"
 	"github.com/apache/incubator-trafficcontrol/lib/go-tc"
 	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/auth"
 	"github.com/basho/riak-go-client"
 	"github.com/jmoiron/sqlx"
 )
 
+type DeliveryServiceTenantInfo struct {
+	ID       int    `db:"id"`
+	XMLID    string `db:"xml_id"`
+	TenantID int    `db:"tenant_id"`
+}
+
+// returns true if the user has tenant access on this deliveryservice
+func (dsInfo DeliveryServiceTenantInfo) IsTenantAuthorized(user auth.CurrentUser, db *sqlx.DB) (bool, error) {
+	return auth.IsResourceAuthorizedToUser(dsInfo.TenantID, user, db)
+}
+
+// tenancy check wrapper for deliveryservice
+func userHasDeliveryServiceTenantAccess(user auth.CurrentUser, dsXmlID string, db *sqlx.DB) (bool, error, int) {
+	dsInfo, err := getDeliveryServiceTenantInfo(dsXmlID, db)
+	if err != nil {
+		if dsInfo == nil {
+			return false, fmt.Errorf("deliveryservice lookup failure: %v", err), http.StatusInternalServerError
+		} else {
+			return false, fmt.Errorf("no such deliveryservice: '%s'", dsXmlID), http.StatusBadRequest
+		}
+	}
+	hasAccess, err := dsInfo.IsTenantAuthorized(user, db)
+	if err != nil {
+		return false, fmt.Errorf("user tenancy check failure: %v", err), http.StatusInternalServerError
+	}
+	if !hasAccess {
+		return false, fmt.Errorf("Access to this resource is not authorized"), http.StatusForbidden
+	}
+	return true, nil, http.StatusOK
+}
+
+// returns tenant information for a deliveryservice
+func getDeliveryServiceTenantInfo(xmlId string, db *sqlx.DB) (*DeliveryServiceTenantInfo, error) {
+	ds := DeliveryServiceTenantInfo{}
+	tenantInfoStmt, err := db.Preparex("SELECT xml_id,tenant_id FROM deliveryservice where xml_id = $1")
+	if err != nil {
+		log.Errorf("deliverservice sql query error, xmlId '%s': %v\n", xmlId, err)
+		return nil, err
+	}
+
+	err = tenantInfoStmt.Get(&ds, xmlId)
+	switch {
+	case err == sql.ErrNoRows:
+		ds = DeliveryServiceTenantInfo{ID: -1, XMLID: "", TenantID: -1}
```

Comment:
does -1 have special meaning here?   should just use the zero value for ds,  so this assignment isn't needed.